### PR TITLE
Fix npgsql/EntityFramework6.Npgsql#52

### PR DESCRIPTION
**Pull request by rherzfeld**
Null string comparison - `Error:could not determine data type of parameter $x`

## Related
#52

### CODE REVIEW

**Review by austindrenski — APPROVED** [review]

**Comment by Emill** [comment]
@roji can you merge this?

**Comment by roji** [comment]
@Emill just to be sure... I don't remember (or ever knew) that much about how EF6 works, so a couple of questions just because this seems a bit risky:

* This would add a `::text` cast for string properties only, within an IS NULL check.
* Doesn't the same problem exists with other reference types, i.e. what happens if someone tries to check `byte[] = null`?
* It's a little odd to cast to text, since this may be executed on non-`text` columns mapped to a string (jsonb/citext, not even sure if this works in EF6). However, as this is done for null checks only, and casting a jsonb/citext to text always succeeds.

Basically I'm ready to merge this, but just want to make sure we've thought about all the possible consequences and avoid breaking people.

**Comment by Emill** [comment]
For all parameters sent, we also send the type, except for string values, which are marked "unknown". Therefore the problem shouldn't exist with `byte[]`.

For "unknown" values, postgresql tries to figure out which type the value has by looking at the context. If it can't figure out a type, this error is thrown.

If it finds a type (for example jsonb) for the parameter, and the parameter is also present in this null check, the value will first be interpreted as a jsonb, then converted to `text` and then compared to `null`.

A little concern I have is if the performance will be worse with this, since postgresql now must also convert the value to `text` before the `null` comparison. But maybe postgresql optimizes this.
